decisions/0006-refactoring-interactions-with-kubernetes-resources.md

```diff
@@ -0,0 +1,86 @@
+# 6. Refactoring Interactions with Kubernetes Resources
+
+Date: 2020-05-11
+
+## Status
+
+Accepted
+
+## Context
+
+The various components deployed via `cloud_controller_ng` interact with the
+Kubernetes API to perform CRUD operations on a variety of Kubernetes Custom
+Resource Definitions (CRDs) which are created in the context of a `cf-for-k8s`
+deployment such that the Kubernetes API becomes the source of truth for those CF
+domain objects (e.g. route resources for apps and build/image resources for app
+builds).
+
+Currently, we have individual clients which interact with a specific Kubernetes
+API group which translates to a client which is scoped to only a particular
+grouping of Kubernetes resources. Currently there are two clients: one to
+interact with `kpack`-related resources and one to to interact with app
+networking-related resources.
+
+We have found that there is an increasing amount of code duplication and shared
+logic/concerns between these two clients and any other such clients we might
+introduce, so we would like to take note of some refactoring ideas to mitigate
+such concerns.
+
+## Decision
+
+The existing clients currently take on two responsibilities which we'd like to
+split out:
+1. Translate an object from the CC database into a Kubernetes resource JSON
+   object
+1. Propagate Kubernetes resource JSON to the Kubernetes API and perform any
+   required error handling
+
+#### Translate an object from the CC database
+
+Proposing creating some sort of translation layer which will take various CC
+domain objects from the CC database and construct their equivalent Kubernetes
+CRs (e.g. CC route object => Route CR).
+
+Seems like this would be difficult to achieve generically, so our first pass can
+probably just create a class for each object which needs to be translated.
+
+#### Propagate the Kubernetes resource JSON
+
+We would like to propose creating a single, generic Kubernetes client which
+accepts JSON form of any Kubernetes resource and simply propagates the desired
+resource to the Kubernetes API. This single, generic client should also continue
+to provide elegant, traceable error handling at least as well as we currently
+do in the existing `route_crd_client` and the `kpack_client`.
```

Comment:
> at least as well

Hopefully better. :) We weren't able to troubleshoot some issues in our environments with the current logging.

https://github.com/cloudfoundry/capi-release/issues/176